src/services/pcn-packetcapture/src/Packetcapture.cpp

```diff
@@ -435,4 +376,65 @@ void Packetcapture::to_timeval(std::chrono::system_clock::duration& d, struct ti
 
     tv.tv_sec  = sec.count();
     tv.tv_usec = std::chrono::duration_cast<std::chrono::microseconds>(d - sec).count();
+}
+
+void Packetcapture::setFilter(const std::string &value) {
+    this->filter = value;
+    logger()->info("Inserted filter: {0}", filter);
+    if (filter == "all") {
+      filterCode = "return pcn_pkt_controller(ctx, md, reason);";
+    }else {
+        memset(&cbpf, 0, sizeof(cbpf));
+        filterCompile(filter, &cbpf);
+    }
+    updateFilter();
+}
+
+std::string Packetcapture::getFilter() {
+  return this->filter;
+}
+
+uint32_t Packetcapture::getSnaplen() {
+  return this->snaplen;
+}
+
+void Packetcapture::setSnaplen(const uint32_t &value) {
+    this->snaplen = value;
+    updateFilter();
+}
+
+void Packetcapture::filterCompile(std::string str, struct sock_fprog * cbpf) {
+    int ret,i;
+    struct bpf_program _bpf;
+    const struct bpf_insn *ins;
+    struct sock_filter *out;
+    int link_type = 1;   //ETHERNET type
+
+    ret = pcap_compile_nopcap(65535, link_type, &_bpf, str.c_str(), 1, 0xffffffff);
```

Comment:
It looks like you're forcing a 65535 snaplen here.
That way you won't be able to change it.

Reply:
In this case, I feel Riccardo is correct. The *classical* eBPF code returns different values based on the snaplen, which is used to tell Linux how many bytes we have to return to user, so in cBPF you have to generate a new filter each time you update the snaplen.
In *eBPF*, the amount of data you return to user does not depend on the BPF code, but has to be enforced in some other way, e.g., by copying only the initial part of the packet to the control plane. So, we do not need to re-generate the (cBPF) code each time the snaplen is changed, and we can stay with a "default" value instead.

Reply:
Thank you for your comment @frisso . So we can let this value as fixed.

then perhaps you're referring to this
https://github.com/polycube-network/polycube/pull/273#discussion_r382865408

Reply:
Yes, correct. Also #273  does not need to be done imho.